### PR TITLE
Implement pop

### PR DESCRIPTION
**Pull request by congma**
The new `pop` method follows the semantics of `dict.pop()`.

Fixes issue #21.

### CODE REVIEW

File: lru.c [hunk 1]
```diff
@@ -409,9 +409,46 @@ LRU_setdefault(LRU *self, PyObject *args)
...
+         * because between the last call of item getter (lru_subscript) and
+         * next call of item setter (lru_ass_sub), the item may have been
+         * removed from the LRU as the effect of arbitrary action in code
+         * elsewhere, thus incurring an error that will not be cleared by us.
+         */
+
+        /* Question: does this matter in the current implementation? */
+        PyObject *e_type, *e_value, *e_traceback;
+
+        PyErr_Fetch(&e_type, &e_value, &e_traceback);
```
Comment (amitdev) [inline lru.c]: Why is this required? Is it for handling updates via multiple threads?

File: lru.c [hunk 2]
```diff
@@ -409,9 +409,46 @@ LRU_setdefault(LRU *self, PyObject *args)
...
+        lru_ass_sub(self, key, NULL);
+	PyErr_Restore(e_type, e_value, e_traceback);
+    } else if (default_obj) {
+        PyErr_Clear();
+        Py_INCREF(default_obj);
+        result = default_obj;
+    }
+
+    return result;
```
Comment (amitdev) [inline lru.c]: I guess `dict.pop` throws a `KeyError` if default is not specified and key is missing. Better to keep same behavior here.

Comment (congma) [reply lru.c]: You're right, it's in the standard library specs:

> If _key_ is in the dictionary, remove it and return its value, else return _default_. If _default_ is not given and _key_ is not in the dictionary, a `KeyError` is raised.

I'll be working on it shortly.

File: lru.c (hunk 1 above)
Comment (congma) [reply lru.c]: Good question, I'm not entirely sure if this is required, but I was playing safe, just in case that the call into the C-extension wasn't atomic from the point of view of the interpreter.

File: lru.c (hunk 2 above)
Comment (congma) [reply lru.c]: I made a mistake reading my own code... The quoted (desired) behaviour is already the current behaviour, and this is tested by the following lines in the test script:

https://github.com/congma/lru-dict/blob/98ec21fcc6fb3722e476aa1b363bb936e2a011ba/test/test_lru.py#L248-L250

File: lru.c (hunk 1 above)
Comment (congma) [reply lru.c]: I removed the unnecessary code here.

**Comment by congma** [comment]
Since the CPython interpreter sees a call into a C extension function as atomic, I removed the code in question guarding the interpreter's exception state when removing an item. The same change is also applied to #27.

**Comment by cool-RR** [comment]
1. Has this change made it into a release? I have 1.1.6 and I'm not seeing it. 
2. Is this guaranteed to take the newest item, the oldest or neither? I'm hoping it's the oldest item, or maybe a keyword argument to choose.
3. Python's `dict` also has popitem, can we have that too please?

**Comment by congma** [comment]
I wrote the part concerning `pop` and `popitem`, and here are my answers:

> 2. Is this guaranteed to take the newest item, the oldest or neither? I'm hoping it's the oldest item, or maybe a keyword argument to choose.

The `pop()` call requires one argument `key`, the key to be popped. This is the same as with Python's `dict.pop`. The key must be specified by the caller, even if it needs not be present (in that case the other argument `default` must be used to specify the value for the missing key). The logic of recency does not apply in this case.

> 3. Python's `dict` also has popitem, can we have that too please?

`popitem` is indeed also implemented and merged (see #27). The behaviour is to pop the least-recent item by default, unless overridden by the optional argument `least_recent=False`, in which case the _most recent_ one is popped.

> 1. Has this change made it into a release? I have 1.1.6 and I'm not seeing it.

Looking at the release history, I don't think the changes have made into a release. The code has been merged though, so you may get it by compiling from the HEAD.

**Comment by cool-RR** [comment]
> I wrote the part concerning `pop` and `popitem`, and here are my answers:
> 
> > 1. Is this guaranteed to take the newest item, the oldest or neither? I'm hoping it's the oldest item, or maybe a keyword argument to choose.
> 
> The `pop()` call requires one argument `key`, the key to be popped. This is the same as with Python's `dict.pop`. The key must be specified by the caller, even if it needs not be present (in that case the other argument `default` must be used to specify the value for the missing key). The logic of recency does not apply in this case.

Awesome, thank you. 

> 
> > 1. Python's `dict` also has popitem, can we have that too please?
> 
> `popitem` is indeed also implemented and merged (see #27). The behaviour is to pop the least-recent item by default, unless overridden by the optional argument `least_recent=False`, in which case the _most recent_ one is popped.

Makes sense. Does it have a race condition though? i.e. if it first checks for the least-recent item, does it have anything to prevent someone else from removing it or modifying it (thus making it not the least-recent item) before it is popped?

**Comment by congma** [comment]
> Does it have a race condition though? i.e. if it first checks for the least-recent item, does it have anything to prevent someone else from removing it or modifying it (thus making it not the least-recent item) before it is popped?

No. The code implementing the `LRU` object does nothing of the sort.

Based on the discussion in #22, the code for `LRU` simply relies on CPython's giant lock to make sure "nobody else" modifies it. However, there's a backdoor. Upon removal of the LRU-element when the cache is filled up, if a callback is set, it is called. The callback can pretty much do anything, including bypassing the CPython GIL. In #32 I'm trying to prevent some of the most egregious causes of inconsistency (and crashes) by locking the callback, but currently the lock is not re-entrant, and this opens many ways to deadlock itself.

If you can help or provide feedback, would you please let me know in #32? Thanks!

**Comment by cool-RR** [comment]
> No. The code implementing the `LRU` object does nothing of the sort.
> 
> Based on the discussion in #22, the code for `LRU` simply relies on CPython's giant lock to make sure "nobody else" modifies it. 

Great, thank you! 

> However, there's a backdoor. Upon removal of the LRU-element when the cache is filled up, if a callback is set, it is called. The callback can pretty much do anything, including bypassing the CPython GIL. In #32 I'm trying to prevent some of the most egregious causes of inconsistency (and crashes) by locking the callback, but currently the lock is not re-entrant, and this opens many ways to deadlock itself.
> 
> If you can help or provide feedback, would you please let me know in #32? Thanks!

The case of the callback is far away from my usage, so I won't volunteer for that, sorry.

**Comment by amitdev** [comment]
Somehow I missed to release this change.  It is done now. - you can checkout the latest version. I should setup auto release based on travis ci when I get time. though.
